token/cli/src/main.rs

```diff
@@ -1712,127 +1666,78 @@ async fn command_gc(
+        let token = token_client_from_config(config, &token_pubkey);
+        let associated_token_account = token.get_associated_token_address(&owner);
         let total_balance: u64 = accounts.values().map(|account| account.0).sum();
 
-        if total_balance > 0 && !accounts.contains_key(&associated_token_account) {
-            // Create the associated token account
-            instructions.push(vec![create_associated_token_account(
-                &config.fee_payer.pubkey(),
-                &owner,
-                &token,
-                &config.program_id,
-            )]);
-            lamports_needed += minimum_balance_for_rent_exemption;
-        }
+        for (address, (amount, decimals, frozen)) in accounts {
+            let is_associated = address == associated_token_account;
 
-        for (address, (amount, decimals, frozen, close_authority)) in accounts {
-            match (
-                address == associated_token_account,
-                close_empty_associated_accounts,
-                total_balance > 0,
-            ) {
-                (true, _, true) => continue, // don't ever close associated token account with amount
-                (true, false, _) => continue, // don't close associated token account if close_empty_associated_accounts isn't set
-                (true, true, false) => println_display(
-                    config,
-                    format!("Closing Account {}", associated_token_account),
-                ),
-                _ => {}
+            // only close the associated account if --close-empty-associated-accounts is provided
+            if is_associated && !close_empty_associated_accounts {
+                continue;
             }
 
-            if frozen {
-                // leave frozen accounts alone
+            // never close the associated account if *any* account carries a balance
+            if is_associated && total_balance > 0 {
                 continue;
             }
 
-            let mut account_instructions = vec![];
+            // dont attempt to close frozen accounts
+            if frozen {
+                continue;
+            }
 
             // Sanity check!
             // we shouldn't ever be here, but if we are here, abort!
-            assert!(amount == 0 || address != associated_token_account);
-
-            if amount > 0 {
-                // Transfer the account balance into the associated token account
-                account_instructions.push(transfer_checked(
-                    &config.program_id,
-                    &address,
-                    &token,
-                    &associated_token_account,
-                    &owner,
-                    &config.multisigner_pubkeys,
-                    amount,
-                    decimals,
-                )?);
-            }
-            // Close the account if config.owner is able to
-            if close_authority == owner {
```

Comment:
neeeevermind, _im_ the one who misunderstood. i was reading the implementation for something unrelated and realized that the token program actually refused the owner as an authority if close authority is set (which strikes me as a bit strange honestly, since the default state of None _doesnt_ mean there is no authority)